### PR TITLE
Push multi-arch images

### PR DESCRIPTION
**Pull request by ddelange**
#### Motivation

Create multi-platform images for `modelmesh-runtime-adapter` images.

#### Modifications

- build for platforms `linux/amd64`, `linux/arm64` (`linux/ppc64le`, `linux/s390x` are not supported by `tnsorflow`)
- utilize Docker build cache for `pip` and `dnf` installs as well as `go build`s
- build on `pull_request` as well as on `push` to catch build breaks during PR review;
   only push images to DockerHub on PR merge (push to `main`)
- upgrade runtime image base from ubi-minimal:8.4 to 8.7
- use `ubi8/go-toolset:1.17` for developer image instead of `ubi8/ubi-minimal:8.4`
- update deprecated `checkout` action
- fix infinite docker inside docker bug when running `make run fmt` inside the developer container

#### Related PRs

- [x] [`modelmesh`](https://github.com/kserve/modelmesh/) (https://github.com/kserve/modelmesh/pull/80)
- [x] [`modelmesh-serving`](https://github.com/kserve/modelmesh-serving/) (https://github.com/kserve/modelmesh-serving/pull/314)
- [x] [`rest-proxy`](https://github.com/kserve/rest-proxy/) (https://github.com/kserve/rest-proxy/pull/24)
- [x] https://github.com/kserve/modelmesh-minio-examples/pull/3

Resolves kserve/modelmesh-serving#162

### CODE REVIEW

File: .github/workflows/build-and-push.yml
Comment (ddelange) [inline .github/workflows/build-and-push.yml:24]: build but don't push on PRs
https://github.com/kserve/rest-proxy/pull/22
```suggestion
```

**Comment by ckadner** [comment]
To avoid some of the arch related errors while installing libraries, we should probably use [ubi8/go-toolset](https://catalog.redhat.com/software/containers/ubi8/go-toolset/5ce8713aac3db925c03774d1?container-tabs=gti) as base image instead of `ubi8/minimal`. Should also help to avoid complexities of building or even just installing go inside the Dockerfile

File: .github/workflows/build-and-push.yml
Comment (ddelange) [inline .github/workflows/build-and-push.yml:64]: same for ppc64le ;( https://pypi.org/project/tensorflow/2.12.0/#files

Comment (ckadner) [reply .github/workflows/build-and-push.yml:64]: Right, thank you, I just noticed your comment. Only `linux/amd64` and `linux/arm64` available

- tensorflow-2.12.0-cp311-cp311-**win_amd64**.whl
- tensorflow-2.12.0-cp311-cp311-manylinux_2_17_x86_64.many**linux**2014_**x86_64**.whl (`linux/amd64`)
- tensorflow-2.12.0-cp311-cp311-manylinux_2_17_aarch64.many**linux**2014_**aarch64**.whl (`linux/arm64`)
- tensorflow-2.12.0-cp311-cp311-macosx_10_15_x86_64.whl

File: Dockerfile
Comment (tjohnson31415) [inline Dockerfile:157]: NIT: assuming that these were only needed for debugging cacheing
```suggestion
    pip install tensorflow
```

File: Dockerfile
Comment (tjohnson31415) [inline Dockerfile:37]: NIT: assuming `pip3 list` was put here for debugging and should be removed?
```suggestion
    pip3 install pre-commit
```

**Review by tjohnson31415 — COMMENTED** [review]
I confirmed that I could build images with binaries for two different CPU archs using the approach coded up here.
 
Cool Stuff!

File: Dockerfile
Comment (ckadner) [reply Dockerfile:130]: I had this and removed it after reading this:

> We could write `FROM --platform=$TARGETPLATFORM` but that is the default value for all builds stages anyway so using a flag is redundant.

https://www.docker.com/blog/faster-multi-platform-builds-dockerfile-cross-compilation-guide/

File: .github/workflows/build-and-push.yml
Comment (ckadner) [reply .github/workflows/build-and-push.yml:10]: Yes, but unit tests are run in a separate workflow already anyhow, and would fail independently, so this reduces time and resource consumption on the GitHub workflow runner. Once a PR is merged, this `build-and-push` workflow will run without test since there is no good way to inform/alert anyone of a failure and via branch protection rules we can assume that only PRs where tests succeed get merged.

File: Dockerfile
Comment (ckadner) [reply Dockerfile:48]: Right, something to consider. I created a follow-up issue here https://github.com/kserve/modelmesh-serving/issues/362

File: Dockerfile
Comment (ckadner) [reply Dockerfile:57]: Good catch!

File: Dockerfile
Comment (ckadner) [reply Dockerfile:157]: Right, but since they don't take up disk space or cost much time overhead I left them in.

File: Dockerfile
Comment (ckadner) [reply Dockerfile]: Removed.

File: Makefile
Comment (ckadner) [reply Makefile]: Thanks. I learned something and reinstated the original line :-)

File: Dockerfile
Comment (ckadner) [reply Dockerfile]: Removed it.

File: Dockerfile
Comment (ckadner) [reply Dockerfile]: Gone.

**Comment by ckadner** [comment]
Thanks @tjohnson31415 -- I updated the PR with the straight-forward changes you suggested and resolved the corresponding conversations.

I left the others as unresolved for you to respond :-)

File: .github/workflows/build-and-push.yml
Comment (ddelange) [reply .github/workflows/build-and-push.yml:10]: additionally, the `push:` key currently evaluates to false on PRs so the registry isn't polluted.

fwiw a pattern I regularly see is a simple RUN statement at the end of the Dockerfile that functions as a smoketest for CI, e.g. the entrypoint is set up properly and has access to whatever was COPY'd. if this RUN statement doesn't write files to disk, it doesn't add to the final image size.

Comment (tjohnson31415) [reply .github/workflows/build-and-push.yml:10]: Thanks for the context and details. Yeah, It seems unlikely that this would actually cause an image to be pushed that doesn't pass tests so long as no one pushes directly to `main`.

Comment (ckadner) [reply .github/workflows/build-and-push.yml:10]: I disabled the ability to push directly to main, even for project owners/maintainers

File: Dockerfile
Comment (tjohnson31415) [inline Dockerfile:35]: So I've been experimenting with this a bit. Does this match your understanding?
- the $HOME for `root` in `go-toolset` is not `/root/` so we need to set the `PIP_CACHE_DIR` env var to configure pip's cache directory
- `ARG` interpolation doesn't work in the `--mount`, so the value is copied instead of using `target=$PIP_CACHE_DIR`

If that's the case, I think it is slightly cleaner to set the PIP_CACHE_DIR env var in the `RUN` command since the `ARG` is not really configurable.

Though I also found https://github.com/moby/buildkit/issues/815 and it sounds like using ARGs in `--mount` can now work, but requires a newer Dockerfile syntax `# syntax=docker/dockerfile:1.3.0`.

File: Makefile
Comment (tjohnson31415) [reply Makefile]: Though the original line also has a problem I think. `$(MAKECMDGOALS)` is the set of goals specified on the cmd-line, but more targets will run if any of those goals have dependencies. So `.PHONY: all $(MAKECMDGOALS)` is only sufficient if no targets in the Makefile have dependencies. If there is a dependency, it will not be .PHONY unless explicitly listed. In this case, `build.develop` should still be in there.

So adding all the targets to .PHONY isn't a bad idea IMO, but there's no reason to add them all and also have `$(MAKECMDGOALS)`.

Makefile hackery 😅

File: Dockerfile
Comment (ddelange) [reply Dockerfile:35]: Hi 👋

Unless the host system (github actions) would build python wheels from source, there is no speedup from mounting pip's cache dir. We generally recommend to [set](https://github.com/alpha-affinity/snakepacker/blob/d1d7a7996fed2d4106e2a567404612b2bfda6340/base/Dockerfile#L24) PIP_NO_CACHE_DIR and rely on docker layer caching by running pip install before copying over source code. That way, you reuse the registry's layer as long as your `requirements.txt` doesn't change. Here's a live multi-arch [example](https://github.com/alpha-affinity/libpostal-fastapi/blob/c200764b9aa8d015ff9d1d5015f42296e238f600/Dockerfile#L21) that combines a compiled C binary (libpostal) with python builds that use it.

Comment (ckadner) [reply Dockerfile:35]: Good comments/questions @tjohnson31415 @ddelange !

> Unless the host system (github actions) would build python wheels from source, there is no speedup from mounting pip's cache dir.

The cache mount is useful for subsequent builds, multi-stage and multi-platform builds. I have seen it being used and speedup the GitHub action builds quite a bit, with significant time savings for installing `grpc` and `tensorflow`

> the $HOME for `root` in `go-toolset` is not `/root/` so we need to set the `PIP_CACHE_DIR` env var to configure pip's cache directory

Yes, otherwise we see this [error](https://github.com/kserve/modelmesh-runtime-adapter/actions/runs/4675184385/jobs/8280043415#step:7:492):
```
#13 12.99 The directory '/opt/app-root/src/.cache/pip/http' or its parent directory is not owned by the current user and the cache has been disabled. Please check the permissions and owner of that directory. If executing pip with sudo, you may want sudo's -H flag.
#13 13.00 The directory '/opt/app-root/src/.cache/pip' or its parent directory is not owned by the current user and caching wheels has been disabled. check the permissions and owner of that directory. If executing pip with sudo, you may want sudo's -H flag.
```

However, for `pip`, the inline `--cache-dir` CLI flag did not set the cache correctly for both HTTP download and the Wheel cache directories. So I use the `ARG PIP_CACHE_DIR=/root/.cache/pip` which works for both.

> `ARG` interpolation doesn't work in the `--mount`, so the value is copied instead of using `target=$PIP_CACHE_DIR`

Yes. I have not tried the new syntax `# syntax=docker/dockerfile:1.3.0` though. That's a good find.

---

References:
- https://www.docker.com/blog/faster-multi-platform-builds-dockerfile-cross-compilation-guide/
- https://github.com/moby/buildkit/blob/master/frontend/dockerfile/docs/reference.md#run---mounttypecache
- https://docs.docker.com/build/cache/
- https://github.com/moby/buildkit/issues/1673

File: Makefile
Comment (ckadner) [reply Makefile]: Since this is tricky to understand -- and apparently was not understood by all developers who made updates to the `Makefile` before ;-) -- I suggest we go with explicitly listing of all goals.
However, in the current form, this still leaves the door wide open for future developers who add new goals (or remove unused ones, or rename existing ones) to overlook/forget to update the .PHONY pseudo target at the end of the Makefile. So I would prefer to “tag” each individual goal like so:

```Makefile
.PHONY: fmt
fmt:
	./scripts/fmt.sh

.PHONY: test
test:
	go test -coverprofile cover.out `go list ./...`
```

This way, the `.PHONY` "tag" would get deleted when the corresponding target gets deleted (or renamed along with renaming a target) and when new targets are added, very likely an existing target would be used as a “template” and the `.PHONY` “tag” would get copy-pasted or added for sheer “consistency” by the uninitiated

Comment (ddelange) [reply Makefile]: Makefiles are fun 😅 I came across a help command that regexes double hashtag above the command, and forms a helptext https://github.com/ddelange/mapply/blob/0.1.21/Makefile

**Review by tjohnson31415 — APPROVED** [review]
LGTM

**Comment by ddelange** [comment]
awesome work @ckadner!!

File: Makefile
Comment (ckadner) [reply Makefile]: @ddelange -- sorry I completely missed your comment. I have done a similar thing in the past as well [here](https://github.com/machine-learning-exchange/mlx/blob/main/Makefile#L9) using `grep` and `awk`

Do you want to open a PR for this, in this and the other 4 ModelMesh repos?

**Pull request merged**
